src/library/autodj/autodjprocessor.cpp

```diff
@@ -1659,3 +1681,67 @@ bool AutoDJProcessor::nextTrackLoaded() {
 
     return loadedTrack == getNextTrackFromQueue();
 }
+
+double AutoDJProcessor::getFadeTime() {
+    double leftBPM = m_pLeftBPM->get();
+    double rightBPM = m_pRightBPM->get();
+    double fromBPM, toBPM;
+
+    double transitionTime = m_transitionTime;
+    double fadeTime;
+
+    double crossfader = getCrossfader();
+    DeckAttributes* pLeftDeck = m_decks[0];
+    DeckAttributes* pRightDeck = m_decks[1];
+
+    if (pLeftDeck->isPlaying() &&
+            (!pRightDeck->isPlaying() || crossfader < 0.0)) {
+        fromBPM = leftBPM;
+        toBPM = rightBPM;
+    } else if (pRightDeck->isPlaying()) {
+        fromBPM = rightBPM;
+        toBPM = leftBPM;
+    } else {
+        // Neither deck is playing. This state isn't reached.
+        // Use this as fallback though just in case.
+        fromBPM = leftBPM;
+        toBPM = rightBPM;
+    }
+
+    if (m_transitionUnit == TransitionUnit::BEATS) {
+        if (leftBPM == rightBPM) {
+            fadeTime = (transitionTime * 60) / leftBPM;
```

Comment:
This assumes that the BPM is constant. @hacksdump and @crisclacerda are working hard to make Mixxx work better with tracks that have changing tempos. Instead of getting the BPM from a ControlProxy, AutoDJProcessor should get the Beats object from the Track object when a Track is loaded and call Beats' method for finding the beats.

Reply:
What if the user changes the BPM using a Deck's RateSlider, will the Beats object from the Track object contain the updated value?

Reply:
That's a great question. I don't think so. So you may need to multiple the positions from Beats by the `rate_ratio` ControlProxy.

Reply:
Thanks, I'll work on that.

Reply:
That would still assume the tempo stays constant over the transition. :/ For AutoDJ, I think that's a good enough hack because it is *Auto*DJ, so the user probably isn't adjusting the tempo as the transition happens. If we ever make AutoDJ automatically sync and fade tempos that will be significantly more complicated.

Reply:
You could connect to `valueChanged` signals of `rate_ratio` and recalculate the transition times.

Reply:
Oh, we already do this in AutoDJProcessor::playerRateChanged. :smile:

Reply:
> Instead of getting the BPM from a ControlProxy, AutoDJProcessor should get the Beats object from the Track object when a Track is loaded and call Beats' method for finding the beats.

Ping. This still needs to be implemented.

Reply:
I'll work on it and hopefully get it working by the end of this day.

Reply:
I hope [commit e02deecce](https://github.com/mixxxdj/mixxx/pull/2746/commits/e02deecce247f59c9f22df9111c7aa59103e5e86) does this well.